### PR TITLE
Added (Dutch) phonenumber recognition

### PR DESCRIPTION
**Pull request by cjfloermans**
Hi Bram,
As agreed I've added a validation for phone numbers.
I've added 3 regular expressions to recognise Dutch format (+31XXXXXXXXX, 0031XXXXXXXXX, 0XXXXXXXXX) and finally if the number is numeric at all. In all these cases the last 9 digits are returned. Otherwise it is False and no search on phone number is done (this also fixes 'double searches' in that field imho).

### CODE REVIEW

**Comment by bramstroker** [comment]
Thanks @cjfloermans.

Seems fine by me.
This removes the possibility to search with telephone numbers starting with a specific string, but this makes no sense to me anyway as the user does not know in which format the value exists in the database.

An improvement would be to search for 3 literal values instead of a LIKE `%...` as for the last case no index can be used, and this gets exponentially problematic when the amount of records in the database increases.

When the user searches for `0612345678` (or any other dutch format), we can take the last 9 characters `612345678`,
and than construct the where to be `billing_telephone IN ('0612345678', '+31612345678', '0031612345678')`.
Now an index can be used and the query should be lightning fast.

**Comment by cjfloermans** [comment]
Hi Bram,
I've implemented your proposal;

First a test is performed if the search term seems to be a Dutch phone number by 3 regular expressions. When this evaluates to true, a literal search is done by those three possibilities. If it evaluates to false, a fallback is in place to search the like '%[LAST 9 DIGITS]' when it's a numeric value, to support searches for other countries as well.
Please see this example for the international format (and the 2 other options 0031[9 DIGITS] and 0[9 DIGITS] gave the same response;
robin_result_phone

File: src/DataProvider/SearchDataProvider.php [hunk 1]
```diff
@@ -169,4 +162,97 @@ protected function getOrders(string $searchTerm): Collection
...
+    /**
+     * This method creates a Filter for phone numbers
+     * If it seems a Dutch Phone number, the term will be parsed in 3 forms and used
+     * as a literal search. Otherwise it will do a like search if the searchterm is a
+     * numeric value.
+     *
+     * @param mixed $searchTerm
+     * @return Filter|null
+     */
+    protected function createPhoneNumberSearchFilter($searchTerm): ?Filter
```
Comment (bramstroker) [inline src/DataProvider/SearchDataProvider.php:175]:
```suggestion
    protected function createPhoneNumberSearchFilter(string $searchTerm): ?Filter
```

File: src/DataProvider/SearchDataProvider.php [hunk 2]
```diff
@@ -169,4 +162,97 @@ protected function getOrders(string $searchTerm): Collection
...
+    }
+
+    /**
+     * This method validates if the searchterm is a numeric value.
+     * If so, it returns the last 9 digits. Otherwise, it returns False
+     *
+     * @param mixed $searchTerm
+     * @return integer|false
+     */
+    protected function parsePhoneNumber($searchTerm)
```
Comment (bramstroker) [inline src/DataProvider/SearchDataProvider.php:206]:
```suggestion
    protected function parsePhoneNumber(string $searchTerm)
```

File: src/DataProvider/SearchDataProvider.php [hunk 3]
```diff
@@ -169,4 +162,97 @@ protected function getOrders(string $searchTerm): Collection
...
+
+    /**
+     * This method performs three tests to validate if the value seems to be a Dutch phone number
+     * The first test is international format with + symbol. The second test is international format
+     * and the third test is local format with preceding 0.
+     *
+     * @param mixed $searchTerm
+     * @return boolean
+     */
+    protected function seemsDutchNumber($searchTerm): bool
```
Comment (bramstroker) [inline src/DataProvider/SearchDataProvider.php:226]:
```suggestion
    protected function seemsDutchNumber(string $searchTerm): bool
```

File: src/DataProvider/SearchDataProvider.php [hunk 4]
```diff
@@ -169,4 +162,97 @@ protected function getOrders(string $searchTerm): Collection
...
+        return false;
+    }
+
+    /**
+     * This method returns an array with formatted phone numbers
+     *
+     * @param mixed $phoneNumber
+     * @return string[]
+     */
+    protected function createDutchPhoneNumbers($phoneNumber): array
```
Comment (bramstroker) [inline src/DataProvider/SearchDataProvider.php:249]:
```suggestion
    protected function createDutchPhoneNumbers(string $phoneNumber): array
```

**Review by bramstroker — CHANGES_REQUESTED** [review]

**Pull request merged**